elrond-wasm/src/storage/mappers/bi_di_mapper.rs

```diff
@@ -0,0 +1,269 @@
+    }
+
+    fn set_id(&mut self, value: &V, id: &K) {
+        storage_set(self.get_id_key(value).as_ref(), id);
+    }
+
+    pub fn get_value(&self, id: &K) -> V {
+        storage_get(self.get_value_key(id).as_ref())
+    }
+
+    fn set_value(&mut self, id: &K, value: &V) {
+        storage_set(self.get_value_key(id).as_ref(), value);
+    }
+
+    pub fn insert(&mut self, id: K, value: V) -> bool {
+        if self.id_set_mapper.contains(&id) || self.value_set_mapper.contains(&value) {
+            return false;
+        }
+        self.set_id(&value, &id);
+        self.set_value(&id, &value);
+
+        self.id_set_mapper.insert(id);
+        self.value_set_mapper.insert(value);
+        true
+    }
+
+    pub fn remove_by_id(&mut self, id: &K) -> bool {
+        if self.id_set_mapper.swap_remove(id) {
+            let value = self.get_value(id);
+            storage_clear(self.get_id_key(&value).as_ref());
+            storage_clear(self.get_value_key(id).as_ref());
+            self.value_set_mapper.swap_remove(&value);
+            return true;
+        }
+        false
+    }
+    pub fn remove_by_value(&mut self, value: &V) -> bool {
+        if self.value_set_mapper.swap_remove(value) {
+            let id = self.get_id(value);
+            storage_clear(self.get_id_key(value).as_ref());
+            storage_clear(self.get_value_key(&id).as_ref());
+            self.id_set_mapper.swap_remove(&id);
+            return true;
+        }
+        false
+    }
+
+    pub fn remove_all_by_ids<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = K>,
+    {
+        for item in iter {
+            self.remove_by_id(&item);
+        }
+    }
+
+    pub fn remove_all_by_values<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = V>,
+    {
+        for item in iter {
+            self.remove_by_value(&item);
+        }
+    }
+
+    pub fn get_all_values(&self) -> ManagedVec<SA, V> {
+        let mut result = ManagedVec::new();
+        for value in self.value_set_mapper.iter() {
+            result.push(value);
+        }
+        result
+    }
+
+    pub fn get_all_ids(&self) -> ManagedVec<SA, K> {
+        let mut result = ManagedVec::new();
+        for id in self.id_set_mapper.iter() {
+            result.push(id);
+        }
+        result
+    }
+
+    pub fn len(&self) -> usize {
+        self.value_set_mapper.len()
+    }
+}
+
+impl<SA, K, V> TopEncodeMulti for BiDiMapper<SA, K, V>
+where
+    SA: StorageMapperApi,
+    K: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+    V: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+{
+    type DecodeAs = MultiValueEncoded<SA, V>;
+
+    fn multi_encode_or_handle_err<O, H>(&self, output: &mut O, h: H) -> Result<(), H::HandledErr>
```

Comment:
I think this should return all (key, value) pairs instead of only the values.

Reply:
Yes, same as for the MapMapper